db/dcrpg/queries.go

```diff
@@ -1287,6 +1287,58 @@ func retrieveAddressTxns(ctx context.Context, db *sql.DB, address string, N, off
 	return scanAddressQueryRows(rows)
 }
 
+// retrieveAddressIoCsv grabs rows for an address and formats them as a 2-D
+// array of strings for CSV-formatting
+func retrieveAddressIoCsv(ctx context.Context, db *sql.DB, address string) (csvRows [][]string, err error) {
+	dbRows, err := db.QueryContext(ctx, internal.SelectAddressCsvView, address)
+	if err != nil {
+		return nil, err
+	}
+	defer closeRows(dbRows)
+
+	var txHash, matchingTxHash, strValidMainchain, strDirection string
+	var validMainchain, isFunding bool
+	var value uint64
+	var ioIndex, txType int
+	var blockTime dbtypes.TimeDef
+
+	// header row
+	csvRows = append(csvRows, []string{"tx_hash", "direction", "io_index", "valid_mainchain", "value", "time_stamp", "tx_type", "matching_tx_hash"})
+
+	for dbRows.Next() {
+
+		err = dbRows.Scan(&txHash, &validMainchain, &matchingTxHash,
+			&value, &blockTime.T, &isFunding, &ioIndex, &txType)
+		if err != nil {
+			return nil, fmt.Errorf("retrieveAddressIoCsv Scan error: %v", err)
+		}
+
+		if validMainchain {
+			strValidMainchain = "1"
+		} else {
+			strValidMainchain = "0"
+		}
+
+		if isFunding {
+			strDirection = "1"
+		} else {
+			strDirection = "-1"
+		}
+
+		csvRows = append(csvRows, []string{
+			txHash,
+			strDirection,
+			fmt.Sprint(ioIndex),
+			strValidMainchain,
+			fmt.Sprintf("%f", dcrutil.Amount(value).ToCoin()),
+			fmt.Sprint(blockTime.T.Unix()),
```

Comment:
These `fmt.Sprint` calls are probably no big deal, but note that `strconv.FormatInt` has a fast path for small integers.

Reply:
`strconv.FormatFloat` may also be worth looking into.  It has optimizations for a bunch of special cases.

```go
FormatFloat(coin, 'f', -1, 64) // -1 precision means use exactly the number if digits required
```

With the `-1` precision argument, it guarantees to return the "smallest number of digits necessary such that ParseFloat will return f exactly".  I don't know if the `fmt` package makes the same guarantee.